RFC-0004-sparse-fill-value/README.md

```diff
@@ -0,0 +1,615 @@
+    the sparse tensor constructors may have different (smaller) shape
+    than the fill value of the hybrid tensor (as defined by point 5).
+
+    The specified fill value (after converting it to `torch.Tensor`
+    instance) can be acquired via `_fill_value()` method.
+
+    For example, the fill value of a (1+1)-D hybrid tensor can be
+    specified as scalar:
+
+    ```python
+    A = torch.sparse_coo_tensor(indices=[[0, 3]],
+                                values=[[.11, .12], [.31, .32]],
+                                size=(4, 2),
+                                fill_value = 1.2)
+    A.fill_value() -> torch.tensor([1.2, 1.2])
+    A._fill_value() -> torch.tensor(1.2)
+    ```
+
+    The output of `fill_value()` is computed as
+
+    ```python
+    A._fill_value().expand(A.values().shape[1:])
+    ```
+
+    Using the specified fill value instead of the fill value of the
+    hybrid tensor as a the internal fill value property, has several
+    advantages:
+
+    - the specified fill value may have a smaller size than the fill
+      value of the hybrid tensor, hence, the memory footprint of a
+      sparse tensor will be slightly smaller,
+    - evaluation of a element-wise function on sparse tensors will be
+      optimal, see point 8 below,
+    - and most importantly, the detection of zero fill value of a
+      hybrid tensor will be optimal, see point 10 below.
+
+7.  The fill value of a sparse tensor can be changed in place.
+
+    For instance:
+
+    ```python
+    A.fill_value().fill_(1.2)
+    ```
+
+    resets the fill value of a sparse tensor `A` to be `1.2`. Note
+    that one can use both `fill_value()` or `_fill_value()` outputs
+    for resetting the fill value of the given tensor.
+
+8.  If `A` is a (coalesced) sparse tensor and `f` is any calculus
+    function that is applied to a tensor element-wise, then:
+
+    ```python
+    f(A) == torch.sparse_coo_tensor(A.indices(), f(A.values()), fill_value=f(A._fill_value()))
+    ```
+
+    From this relation follows an identity:
+
+    ```python
+    f(A).to_dense() == f(A.to_dense())
+    ```
+
+    that will be advantageous in testing the sparse tensor support of
+    element-wise functions.
+
+9.  The fill value of an element-wise n-ary operation on sparse
+    tensors with different fill values is equal to the result of
+    applying the operation to the fill values of the sparse
+    tensors.
+
+    For instance (`*` represents unspecified element),
+
+    ```
+    A = [[1, *],
+         [3, *]]
+    B = [[5, *],
+         [*, 8]]
+    # assume A and B fill values are 2 and 6, respectively, then
+    A + B = [[6,  *],
+             [9, 10]]
+    # with fill value 2 + 6 = 8
+    ```
+
+10. Existing PyTorch functions that support sparse tensor as inputs,
+    need to be updated for handling the defined fill values. This will
+    be implemented in two stages:
+
+    1. All relevant functions need to check for zero fill value. If a
```

Comment:
This is trouble, if fill values live on device. If fill value lives on device, you cannot conveniently check if it is nonzero without inducing a synchronize. Indeed, the pseudocode below's use of `nonzero` would cause a sync. This is bad.

It might be possible to do some conservative analysis; for example, we can separately manage if something is "definitely zero" versus "maybe not zero" on a cpu side bit, and then have individual functions propagate it. But at that point, you might as well just upfront fix all the functions to handle fill value. Fortunately, there aren't that many; it's small enough that, for example, when I did the port from THS to ATen, I did it all in one go.

Reply:
@ezyang Having the fast "definitely zero" test would be advantageous even after fixing all the functions to handle fill value. This is because most Linear Algebra functions would always do the zero-test in order to use the fastest and most frequently used zero-fill-value-pathway.

An idea: let the fill value always live in the CPU memory so that the zero-test will be fast. When an algorithm needs to use fill value from a device, the algorithm makes the to-device copy.

To make coding easier, one could use unified CUDA memory only for the fill value when the sparse tensor `values` uses CUDA device. Although, other non-CUDA devices would not gain from this and explicit copy-fill-value-to-device would be still needed.